README.md
co-urllib
=========

co version of [urllib](https://github.com/fengmk2/urllib).

Not just thunkify, `co-urllib.request` is a [Generator](http://wiki.ecmascript.org/doku.php?id=harmony:generators).

[![NPM](https://nodei.co/npm/co-urllib.png?downloads=true)](https://nodei.co/npm/co-urllib/)

Code coverage: [100%](http://qtestbucket.qiniudn.com/cov/html/co-urllib/0.0.1/index.html)

## Usage

```
var urllib = require('co-urllib');
var co = require('co');

co(function *() {
  var result = yield urllib.request('http://baidu.com');
  var data = result.data; // response data
  var status = result.status; // response status code
  var headers = result.headers; // response headers
  console.log(status);
})();
```

More documents, please @see [API of urllib](https://github.com/fengmk2/urllib#api-doc)

## Licences

(The MIT License)

Copyright (c) 2013 - 2014 dead-horse and other contributors

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

0.2.2 / 2014-05-12
==================

 * add current request url to err.message

0.2.1 / 2014-04-11
==================

  * remove res.aborted

0.2.0 / 2014-04-08
==================

 * use ali.regenerator to skip download esprima from github
 * use regenerator, detect generator support

0.1.4 / 2014-03-26
==================

 * update deps
 * use co-mocha

0.1.3 / 2014-03-14
==================

  * exports default USER_AGENT

0.1.2 / 2014-03-13
==================

  * JSONResponseFormatError should contains raw data

0.1.1 / 2014-03-13
==================

  * add status and headers to error

0.1.0 / 2014-03-13
==================

  * update test
  * Merge pull request #2 from dead-horse/co
  * support args.writeStream and run test on travis-ci
  * add more test cases
  * Refactor request using generator. fixed #1

0.0.1 / 2013-12-31
==================

  * update readme
  * add co urllib code
  * Initial commit


test/server.js
/**!
 * co-urllib - server.js
 *
 * Copyright(c) fengmk2 and other contributors.
 * MIT Licensed
 *
 * Authors:
 *   fengmk2 <fengmk2@gmail.com> (http://fengmk2.github.com)
 */

'use strict';

/**
 * Module dependencies.
 */

var koa = require('koa');
var sleep = require('co-sleep');
var middlewares = require('koa-middlewares');
var speeds = require('speeds');
var fs = require('fs');

var app = koa();
app.use(middlewares.bodyParser());

app.use(function *(next) {
  if (this.url === '/sleep') {
    yield sleep(1000);
  } else if (this.url === '/wrongjson') {
    return this.body = '{foo:bar}';
  } else if (this.url === '/json') {
    return this.body = {foo: 'bar'};
  } else if (this.url === '/destroy') {
    return this.res.destroy();
  } else if (this.url.indexOf('/qs') === 0) {
    return this.body = this.query;
  } else if (this.url.indexOf('/post') === 0) {
    return this.body = {
      body: this.request.body,
      'content-type': this.get('content-type'),
    };
  } else if (this.url === '/ua') {
    return this.body = {ua: this.get('user-agent')};
  } else if (this.url === '/direct') {
    return this.body = this.req;
  } else if (this.url === '/auth') {
    var auth = new Buffer(this.get('authorization').split(' ')[1], 'base64').toString().split(':');
    return this.body = {user: auth[0], pass: auth[1]};
  } else if (this.url === '/302') {
    this.status = 302;
    return this.redirect('/204');
  } else if (this.url === '/301') {
    this.status = 301;
    return this.redirect('/204');
  } else if (this.url === '/redirect_no_location') {
    this.status = 302;
    return this.body = 'I am 302 body';
  } else if (this.url === '/204') {
    this.status = 204;
    return;
  } else if (this.url === '/loop_redirect') {
    this.status = 302;
    return this.redirect('/loop_redirect');
  } else if (this.url === '/content-encoding') {
    this.set('content-encoding', 'foo');
    return this.body = new Buffer('bar');
  } else if (this.url === '/slow') {
    return this.body = fs.createReadStream(__filename).pipe(speeds(100));
  } else if (this.url === '/res-connection-end') {
    var res = this.res;
    setTimeout(function () {
      res.connection.end();
    }, 2000);
    return this.body = fs.createReadStream(__filename).pipe(speeds(100));
  } else if (this.url === '/socket.destroy') {
    var res = this.res;
    setTimeout(function () {
      res.destroy();
    }, 2000);
    return this.body = fs.createReadStream(__filename).pipe(speeds(100));
  }
});

module.exports = app;


test/index.test.js
/**!
 * co-urllib - test/index.js
 *
 * Copyright(c) fengmk2 and other contributors.
 * MIT Licensed
 *
 * Authors:
 *   fengmk2 <fengmk2@gmail.com> (http://fengmk2.github.com)
 */

'use strict';

/**
 * Module dependencies.
 */

var thunkify = require('thunkify');
var should = require('should');
var fs = require('fs');
var gunzip = thunkify(require('zlib').gunzip);
var urllib = require('../');
var app = require('./server');

describe('index.test.js', function () {
  var host = 'http://127.0.0.1:';
  var port = null;

  before(function (done) {
    app = app.listen(0, function () {
      host += app.address().port;
      done();
    });
  });

  describe('USER_AGENT', function () {
    it('should exports USER_AGENT', function () {
      urllib.USER_AGENT.should.be.a.String;
    });
  });

  describe('RequestError', function () {
    it('should return RequestError when server destroy connection', function *() {
      try {
        var result = yield *urllib.request(host + '/destroy', {dataType: 'json'});
        throw new Error('should not run this');
      } catch (e) {
        e.name.should.equal('RequestError');
        e.message.should.include('socket hang up (GET http://127.0.0.1:');
        e.status.should.equal(-1);
        e.headers.should.eql({});
      }
    });

    it('should return RequestError when domain not exists', function *() {
      try {
        var result = yield *urllib.request('http://foo.co-urllib.com', {dataType: 'json'});
        throw new Error('should not run this');
      } catch (e) {
        e.name.should.equal('RequestError');
        e.message.should.equal('getaddrinfo ENOTFOUND (GET http://foo.co-urllib.com)');
        e.status.should.equal(-1);
        e.headers.should.eql({});
      }
    });
  });

  describe('options.beforeRequest', function () {
    it('should change headers before send', function *() {
      var result = yield *urllib.request(host + '/ua', {
        dataType: 'json',
        beforeRequest: function (options) {
          options.headers['User-Agent'] = 'fooagent';
        }
      });
      result.should.have.keys('data', 'status', 'headers');
      result.status.should.equal(200);
      result.data.should.be.an.Object;
      result.data.should.eql({ua: 'fooagent'});
      result.headers['content-type'].should.equal('application/json');
    });
  });

  describe('options.agent = false, httpsAgent = false', function () {
    it('should work with http', function *() {
      var result = yield *urllib.request(host + '/json', {dataType: 'json', agent: false});
      result.should.have.keys('data', 'status', 'headers');
      result.status.should.equal(200);
      result.data.should.be.an.Object;
      result.data.should.eql({foo: 'bar'});
      result.headers['content-type'].should.equal('application/json');
    });

    it('should work with https', function *() {
      var result = yield *urllib.request('https://npmjs.org', {httpsAgent: false, timeout: 10000});
      result.should.have.keys('data', 'status', 'headers');
      result.data.should.be.a.Buffer;
    });
  });

  describe('header["user-agent"]', function () {
    it('should got default user-agent', function *() {
      var result = yield *urllib.request(host + '/ua', {
        dataType: 'json',
      });
      result.should.have.keys('data', 'status', 'headers');
      result.status.should.equal(200);
      result.data.should.be.an.Object;
      result.data.should.match({ua: /^node-co-urllib\/\d+\.\d+\.\d+ node\//});
      result.headers['content-type'].should.equal('application/json');
    });

    it('should custom user-agent', function *() {
      var result = yield *urllib.request(host + '/ua', {
        dataType: 'json',
        headers: {
          'user-agent': 'testua'
        }
      });
      result.should.have.keys('data', 'status', 'headers');
      result.status.should.equal(200);
      result.data.should.be.an.Object;
      result.data.should.eql({ua: 'testua'});
      result.headers['content-type'].should.equal('application/json');
    });
  });

  describe('options.rejectUnauthorized = ', function () {
    it('should ignore ssl check', function *() {
      var result = yield *urllib.request('https://npmjs.org', {
        timeout: 10000,
        rejectUnauthorized: false,
      });
      result.should.have.keys('data', 'status', 'headers');
      // result.status.should.equal(200);
      result.data.should.be.a.Buffer;
    });
  });

  describe('options.auth = ', function () {
    it('should send with auth', function *() {
      var result = yield *urllib.request(host + '/auth', {
        dataType: 'json',
        auth: 'foo:bar'
      });
      result.should.have.keys('data', 'status', 'headers');
      result.status.should.equal(200);
      result.data.should.be.an.Object;
      result.data.should.eql({user: 'foo', pass: 'bar'});
      result.headers['content-type'].should.equal('application/json');
    });
  });

  describe('options.writeStream = ', function () {
    it('should save data to stream', function *() {
      var result = yield *urllib.request(host + '/json', {
        writeStream: fs.createWriteStream(__filename + '.out'),
      });
      result.should.have.keys('data', 'status', 'headers');
      result.status.should.equal(200);
      should.not.exist(result.data);
      fs.readFileSync(__filename + '.out', 'utf8').should.equal('{"foo":"bar"}');
    });
  });

  describe('options.stream = ', function () {
    it('should send with stream', function *() {
      var result = yield *urllib.request(host + '/direct', {
        method: 'post',
        stream: fs.createReadStream(__filename),
      });
      result.should.have.keys('data', 'status', 'headers');
      result.status.should.equal(200);
      result.data.should.be.a.Buffer;
      result.data.toString().should.equal(fs.readFileSync(__filename, 'utf8'));
      result.headers['content-type'].should.equal('application/octet-stream');
    });
  });

  describe('options.content = ', function () {
    it('should send with buffer content', function *() {
      var result = yield *urllib.request(host + '/direct', {
        method: 'post',
        dataType: 'json',
        content: new Buffer(JSON.stringify({nick: '苏千'})),
      });
      result.should.have.keys('data', 'status', 'headers');
      result.status.should.equal(200);
      result.data.should.be.an.Object;
      result.data.should.eql({nick: '苏千'});
      result.headers['content-type'].should.equal('application/octet-stream');
    });
  });

  describe('options.followRedirect = true', function () {
    it('should follow 302 redirect', function *() {
      var result = yield *urllib.request(host + '/302', {
        followRedirect: true,
      });
      result.should.have.keys('data', 'status', 'headers');
      result.status.should.equal(204);
      should.not.exist(result.data);
    });

    it('should follow 301 redirect', function *() {
      var result = yield *urllib.request(host + '/301', {
        followRedirect: true,
      });
      result.should.have.keys('data', 'status', 'headers');
      result.status.should.equal(204);
      should.not.exist(result.data);
    });

    it('should throw FollowRedirectError', function *() {
      try {
        var result = yield *urllib.request(host + '/redirect_no_location', {
          followRedirect: true,
        });
        throw new Error('should not run this');
      } catch(e) {
        e.name.should.equal('FollowRedirectError');
        e.message.should.include('Got statusCode 302 but cannot resolve next location from headers (GET http://127.0.0.1:');
        e.status.should.equal(302);
        e.should.have.property('headers');
      }
    });

    it('should throw MaxRedirectError with default 10 max', function *() {
      try {
        var result = yield *urllib.request(host + '/loop_redirect', {
          followRedirect: true,
        });
        throw new Error('should not run this');
      } catch(e) {
        e.name.should.equal('MaxRedirectError');
        e.message.should.include('Exceeded 10 maxRedirects. Probably stuck in a redirect loop');
        e.message.should.include('(GET http://127.0.0.1:');
        e.status.should.equal(302);
        e.headers.location.should.equal('/loop_redirect');
      }
    });

    it('should throw MaxRedirectError with default 1 max', function *() {
      try {
        var result = yield *urllib.request(host + '/loop_redirect', {
          followRedirect: true,
          maxRedirects: 1,
        });
        throw new Error('should not run this');
      } catch(e) {
        e.name.should.equal('MaxRedirectError');
        e.message.should.include('Exceeded 1 maxRedirects. Probably stuck in a redirect loop');
        e.status.should.equal(302);
        e.headers.location.should.equal('/loop_redirect');
      }
    });
  });

  describe('options.data = ', function () {
    it('should get with querystring 1', function *() {
      var result = yield *urllib.request(host + '/qs?k=name', {dataType: 'json', data: {nick: '苏千'}});
      result.should.have.keys('data', 'status', 'headers');
      result.status.should.equal(200);
      result.data.should.be.an.Object;
      result.data.should.eql({nick: '苏千', k: 'name'});
      result.headers['content-type'].should.equal('application/json');
    });

    it('should get with querystring 2', function *() {
      var result = yield *urllib.request(host + '/qs', {dataType: 'json', data: {nick: '苏千'}});
      result.should.have.keys('data', 'status', 'headers');
      result.status.should.equal(200);
      result.data.should.be.an.Object;
      result.data.should.eql({nick: '苏千'});
      result.headers['content-type'].should.equal('application/json');
    });

    it('should post with form content', function *() {
      var result = yield *urllib.request(host + '/post',
        {dataType: 'json', data: {nick: '苏千'}, method: 'POST'});
      result.should.have.keys('data', 'status', 'headers');
      result.status.should.equal(200);
      result.data.should.be.an.Object;
      result.data.should.eql({
        body: {nick: '苏千'},
        "content-type": "application/x-www-form-urlencoded"
      });
      result.headers['content-type'].should.equal('application/json');
    });

    it('should post with json content', function *() {
      var result = yield *urllib.request(host + '/post',
        {
          dataType: 'json', data: {nick: '苏千'}, method: 'post',
          headers: {'content-type': 'application/json'}
        }
      );
      result.should.have.keys('data', 'status', 'headers');
      result.status.should.equal(200);
      result.data.should.be.an.Object;
      result.data.should.eql({
        body: {nick: '苏千'},
        "content-type": "application/json"
      });
      result.headers['content-type'].should.equal('application/json');
    });
  });

  describe('options.dataType = json', function () {
    it('should got json', function *() {
      var result = yield *urllib.request(host + '/json', {dataType: 'json'});
      result.should.have.keys('data', 'status', 'headers');
      result.status.should.equal(200);
      result.data.should.be.an.Object;
      result.data.should.eql({foo: 'bar'});
      result.headers['content-type'].should.equal('application/json');
    });

    it('should throw JSONResponseFormatError', function *() {
      try {
        var result = yield *urllib.request(host + '/wrongjson', {dataType: 'json'});
        throw new Error('should not run this');
      } catch (e) {
        e.name.should.equal('JSONResponseFormatError');
        e.status.should.equal(200);
        e.headers['content-type'].should.equal('text/plain; charset=utf-8');
        e.data.should.be.a.Buffer;
        e.data.toString().should.equal('{foo:bar}');
      }
    });
  });

  describe('options.timeout', function () {
    it('should throw ConnectionTimeoutError', function *() {
      try {
        var result = yield *urllib.request(host + '/sleep', {timeout: 10});
        throw new Error('should not run this');
      } catch (e) {
        e.status.should.equal(408);
        e.name.should.equal('ConnectionTimeoutError');
        e.message.should.include('timeout of 10ms exceeded (GET http://127.0.0.1:');
        e.headers.should.eql({});
      }
    });

    it('should throw ResponseTimeoutError', function *() {
      try {
        var result = yield *urllib.request(host + '/slow', {timeout: 2000});
        throw new Error('should not run this');
      } catch (e) {
        should.exist(e);
        e.status.should.equal(408);
        e.name.should.equal('ResponseTimeoutError');
        e.message.should.include('timeout of 2000ms exceeded (GET http://127.0.0.1:');
        e.headers['content-type'].should.equal('application/octet-stream');
      }
    });
  });

  describe('options.gzip = true', function () {
    var gzipurl = 'http://qtestbucket.qiniudn.com/dist/v0.10.28/SHASUMS.txt';
    it('should get gzip response and auto decode it', function *() {
      var result = yield *urllib.request(gzipurl, {
        // dataType: 'json',
        gzip: true,
        timeout: 10000,
      });
      result.should.have.keys('data', 'status', 'headers');
      result.status.should.equal(200);
      result.data.toString().should.include('6eff580cc8460741155d42ef1ef537961194443f');
      // result.data.should.be.a.Object;
      // result.data.name.should.equal('byte');
      // result.headers['content-type'].should.equal('application/json');
      result.headers['content-encoding'].should.equal('gzip');
    });

    it('should get gzip response and decode custom', function *() {
      var result = yield *urllib.request(gzipurl, {
        // dataType: 'json',
        gzip: true,
        timeout: 10000,
        headers: {
          'accept-encoding': 'gzip'
        }
      });
      result.should.have.keys('data', 'status', 'headers');
      result.status.should.equal(200);
      result.data.should.be.an.Buffer;
      result.data.length.should.above(0);
      // result.headers['content-type'].should.equal('application/json');
      result.headers['content-encoding'].should.equal('gzip');
      var buf = yield gunzip(result.data);
      buf.toString().should.include('6eff580cc8460741155d42ef1ef537961194443f');
      // JSON.parse(buf).name.should.equal('byte');
    });

    it('should get response with content-encoding', function *() {
      var result = yield *urllib.request(host + '/content-encoding', {
        dataType: 'json',
        gzip: true,
      });
      result.should.have.keys('data', 'status', 'headers');
      result.status.should.equal(200);
      result.data.should.be.an.Buffer;
      result.data.length.should.above(0);
      result.headers['content-encoding'].should.equal('foo');
      result.data.toString().should.equal('bar');
    });
  });

  describe('http and https web site', function () {
    it('should get http://nodejs.org success', function *() {
      var result = yield *urllib.request('http://nodejs.org', {timeout: 10000});
      result.should.have.keys('data', 'status', 'headers');
      result.status.should.equal(200);
      result.data.should.be.a.Buffer;
      result.data.length.should.above(0);
      result.headers['content-type'].should.equal('text/html');
    });

    it('should get https://registry.npmjs.org success', function *() {
      var result = yield *urllib.request('https://registry.npmjs.org', {dataType: 'json', timeout: 10000});
      result.should.have.keys('data', 'status', 'headers');
      result.status.should.equal(200);
      result.data.should.be.an.Object;
      result.data.db_name.should.equal('registry');
    });
  });
});


test/callback.test.js
/**!
 * co-urllib - test/callback.test.js
 *
 * Copyright(c) fengmk2 and other contributors.
 * MIT Licensed
 *
 * Authors:
 *   fengmk2 <fengmk2@gmail.com> (http://fengmk2.github.com)
 */

'use strict';

/**
 * Module dependencies.
 */

var co = require('co');
var should = require('should');
var urllib = require('../build/urllib');

var request = co(urllib.request);

describe('callback.test.js', function () {
  it('should get http://nodejs.org success', function (done) {
    request('http://nodejs.org', {timeout: 10000}, function (err, result) {
      should.not.exist(err);
      result.should.have.keys('data', 'status', 'headers');
      result.status.should.equal(200);
      result.data.should.be.a.Buffer;
      result.data.length.should.above(0);
      if (result.headers['content-length']) {
        result.data.should.length(+result.headers['content-length']);
      }
      result.headers['content-type'].should.equal('text/html');
      done();
    });
  });
});


index.js
/**!
 * co-urllib - index.js
 *
 * MIT Licensed
 *
 * Authors:
 *   dead_horse <dead_horse@qq.com> (http://deadhorse.me)
 *   fengmk2 <fengmk2@gmail.com> (http://fengmk2.cnpmjs.org)
 */

'use strict';

/**
 * Module dependencies.
 */

var debug = require('debug')('co-urllib');
var thunkify = require('thunkify');
var http = require('http');
var https = require('https');
var urlutil = require('url');
var qs = require('querystring');
var path = require('path');
var fs = require('fs');
var zlib = require('zlib');
var readall = require('co-readall');
var assertTimeout = require('co-assert-timeout');
var ua = require('default-user-agent');
var gunzip = thunkify(zlib.gunzip);

var pkg = JSON.parse(fs.readFileSync(path.join(__dirname, 'package.json')));

var REQUEST_ID = 0;

function createRequest(httplib) {
  return function (options, args) {
    var reqId = ++REQUEST_ID;
    return function (done) {
      var called = false;
      function _done(err, result) {
        if (called) {
          return;
        }
        called = true;
        done(err, result);
      }

      var req = httplib.request(options, function (res) {
        _done(null, {req: req, res: res});
      });

      req.requestId = reqId;

      req.on('error', function onerror(err) {
        if (err.name === 'Error') {
          err.name = 'RequestError';
        }
        debug('Request#%d %s `req error` event emit, %s: %s', reqId, options.path, err.name, err.message);
        _done(err);
      });

      if (args.stream) {
        args.stream.pipe(req);
      } else {
        req.end(args.body);
      }
    };
  };
}

var httpRequest = createRequest(http);
var httpsRequest = createRequest(https);

var USER_AGENT = exports.USER_AGENT = ua('node-co-urllib', pkg.version);

// change Agent.maxSockets to 1000
exports.agent = new http.Agent();
exports.agent.maxSockets = 1000;

exports.httpsAgent = new https.Agent();
exports.httpsAgent.maxSockets = 1000;

/**
 * The default request timeout(in milliseconds) 5000ms.
 * @type {Number}
 * @const
 */
exports.TIMEOUT = 5000;

/**
 * Handle all http request, both http and https support well.
 *
 * @example
 *
 * // GET http://httptest.cnodejs.net
 * var result = yield *urllib.request('http://httptest.cnodejs.net/test/get');
 * // POST http://httptest.cnodejs.net
 * var args = { type: 'post', data: { foo: 'bar' } };
 * var result = yield *urllib.request('http://httptest.cnodejs.net/test/post', args);
 *
 * @param {String|Object} url
 * @param {Object} [args], optional
 *   - {Object} [data]: request data, will auto be query stringify.
 *   - {String|Buffer} [content]: optional, if set content, `data` will ignore.
 *   - {ReadStream} [stream]: read stream to sent.
 *   - {WriteStream} [writeStream]: writable stream to save response data.
 *       If you use this, callback's data should be null.
 *       We will just `pipe(ws, {end: true})`.
 *   - {String} [method]: optional, could be GET | POST | DELETE | PUT, default is GET
 *   - {String} [dataType]: optional, `text` or `json`, default is text
 *   - {Object} [headers]: optional, request headers
 *   - {Number} [timeout]: request timeout(in milliseconds), default is `exports.TIMEOUT`
 *   - {Agent} [agent]: optional, http agent. Set `false` if you does not use agent.
 *   - {Agent} [httpsAgent]: optional, https agent. Set `false` if you does not use agent.
 *   - {String} [auth]: Basic authentication i.e. 'user:password' to compute an Authorization header.
 *   - {String|Buffer|Array} [ca]: An array of strings or Buffers of trusted certificates.
 *       If this is omitted several well known "root" CAs will be used, like VeriSign.
 *       These are used to authorize connections.
 *       Notes: This is necessary only if the server uses the self-signed certificate
 *   - {Boolean} [rejectUnauthorized]: If true, the server certificate is verified against the list of supplied CAs.
 *       An 'error' event is emitted if verification fails. Default: true.
 *   - {String|Buffer} [pfx]: A string or Buffer containing the private key,
 *       certificate and CA certs of the server in PFX or PKCS12 format.
 *   - {String|Buffer} [key]: A string or Buffer containing the private key of the client in PEM format.
 *       Notes: This is necessary only if using the client certificate authentication
 *   - {String|Buffer} [cert]: A string or Buffer containing the certificate key of the client in PEM format.
 *       Notes: This is necessary only if using the client certificate authentication
 *   - {String} [passphrase]: A string of passphrase for the private key or pfx.
 *   - {Boolean} [followRedirect]: Follow HTTP 3xx responses as redirects. defaults to false.
 *   - {Number} [maxRedirects]: The maximum number of redirects to follow, defaults to 10.
 *   - {Function(options)} [beforeRequest]: Before request hook, you can change every thing here.
 *   - {Boolean} [gzip]: Accept gzip response content and auto decode it, default is `false`.
 * @return {Object} result, contains `data` and `res`
 *   - {Buffer|Object} data
 *   - {Response} res
 * @api public
 */
function *request(url, args) {
  args = args || {};

  args.timeout = args.timeout || exports.TIMEOUT;
  args.maxRedirects = args.maxRedirects || 10;
  var parsedUrl = typeof url === 'string' ? urlutil.parse(url) : url;

  var method = (args.type || args.method || parsedUrl.method || 'GET').toUpperCase();
  var port = parsedUrl.port || 80;
  var _request = httpRequest;
  var agent = args.agent || exports.agent;

  if (parsedUrl.protocol === 'https:') {
    _request = httpsRequest;
    agent = args.httpsAgent || exports.httpsAgent;
    if (args.httpsAgent === false) {
      agent = false;
    }
    if (!parsedUrl.port) {
      port = 443;
    }
  }

  if (args.agent === false) {
    agent = false;
  }

  var options = {
    host: parsedUrl.hostname || parsedUrl.host || 'localhost',
    path: parsedUrl.path || '/',
    method: method,
    port: port,
    agent: agent,
    headers: args.headers || {}
  };

  var sslNames = ['ca', 'pfx', 'key', 'cert', 'passphrase'];
  for (var i = 0; i < sslNames.length; i++) {
    var name = sslNames[i];
    if (args[name]) {
      options[name] = args[name];
    }
  }

  if (args.rejectUnauthorized !== undefined) {
    options.rejectUnauthorized = args.rejectUnauthorized;
  }

  var auth = args.auth || parsedUrl.auth;
  if (auth) {
    options.auth = auth;
  }

  var body = args.content || args.data;
  var isReadAction = method === 'GET' || method === 'HEAD';
  if (!args.content) {
    if (body && !(typeof body === 'string' || Buffer.isBuffer(body))) {
      if (isReadAction) {
        // read: GET, HEAD, use query string
        body = qs.stringify(body);
      } else {
        // auto add application/x-www-form-urlencoded when using urlencode form request
        if (!options.headers['Content-Type'] && !options.headers['content-type']) {
          options.headers['Content-Type'] = 'application/x-www-form-urlencoded';
        }

        var contentType = options.headers['Content-Type'] || options.headers['content-type'];
        if (contentType === 'application/json') {
          body = JSON.stringify(body);
        } else {
          // 'application/x-www-form-urlencoded'
          body = qs.stringify(body);
        }
      }
    }
  }

  // if it's a GET or HEAD request, data should be sent as query string
  if (isReadAction && body) {
    options.path += (parsedUrl.query ? '&' : '?') + body;
    body = null;
  }

  if (body) {
    var length = body.length;
    if (!Buffer.isBuffer(body)) {
      length = Buffer.byteLength(body);
    }
    options.headers['Content-Length'] = length;
  }

  args.dataType = args.dataType || args.datatype;

  if (args.dataType === 'json') {
    options.headers.Accept = 'application/json';
  }

  if (typeof args.beforeRequest === 'function') {
    // you can use this hook to change every thing.
    args.beforeRequest(options);
  }

  // set user-agent
  if (!options.headers['User-Agent'] && !options.headers['user-agent']) {
    options.headers['User-Agent'] = USER_AGENT;
  }

  var enableGzip = args.gzip === true;
  if (enableGzip) {
    var acceptEncoding = options.headers['Accept-Encoding'] || options.headers['accept-encoding'];
    if (acceptEncoding) {
      enableGzip = false; // user want to handle response content decode themself
    } else {
      options.headers['Accept-Encoding'] = 'gzip';
    }
  }

  debug('%s %s, headers: %j', method, url, options.headers);

  var r;
  try {
    r = yield assertTimeout(_request(options, {body: body, stream: args.stream}), args.timeout);
  } catch (err) {
    err.status = err.status || -1;
    if (err.status === 408) {
      err.name = 'ConnectionTimeoutError';
    }
    err.headers = {};
    throw err;
  }

  var req = r.req;
  var res = r.res;

  var reqId = req.requestId;

  debug('Request#%d %s `req response` event emit: status %d, headers: %j',
    reqId, options.path, res.statusCode, res.headers);

  if ((res.statusCode === 302 || res.statusCode === 301) && args.followRedirect) {  // handle redirect
    args._followRedirectCount = (args._followRedirectCount || 0) + 1;
    var err = null;
    if (!res.headers.location) {
      err = new Error('Got statusCode ' + res.statusCode + ' but cannot resolve next location from headers');
      err.name = 'FollowRedirectError';
    } else if (args._followRedirectCount > args.maxRedirects) {
      err = new Error('Exceeded ' + args.maxRedirects + ' maxRedirects. Probably stuck in a redirect loop ' + url);
      err.name = 'MaxRedirectError';
    }

    if (err) {
      err.status = res.statusCode;
      err.headers = res.headers;
      throw err;
    }

    var _url = urlutil.resolve(url, res.headers.location);
    debug('Request#%d %s: `redirected` from %s to %s', reqId, options.path, url, _url);
    return yield *request(_url, args);
  }

  var aborted = false;
  res.on('aborted', function () {
    aborted = true;
  });

  req.on('close', function () {
    debug('Request#%d %s: `req close` event emit', reqId, options.path);
  });

  var data;
  try {
    data = yield assertTimeout(readall(res, args.writeStream), args.timeout);
  } catch (err) {
    err.requestId = reqId;
    err.status = err.status || res.statusCode;
    if (err.status === 408) {
      req.abort(); // try to abort response handle
      err.name = 'ResponseTimeoutError';
    }
    err.headers = res.headers;
    throw err;
  }

  if (aborted) {
    var err = new Error('Remote socket was terminated before `response.end()` was called');
    err.name = 'RemoteSocketClosedError';
    err.status = res.statusCode;
    err.headers = res.headers;
    throw err;
  }

  var size = data && data.length || 0;

  debug('Request#%d %s %s %s got %d bytes body',
    reqId, method, url, res.statusCode, size);

  var encoding = res.headers['content-encoding'];

  if (size > 0) {
    if (enableGzip) {
      if (encoding && encoding.toLowerCase() === 'gzip') {
        var gzipLength = data.length;
        data = yield gunzip(data);
        encoding = null;
        debug('gunzip %d bytes body to %d bytes', gzipLength, data.length);
      }
    }

    // if response body not decode, dont touch it
    if (!encoding && args.dataType === 'json') {
      try {
        data = JSON.parse(data);
      } catch (err) {
        err.name = 'JSONResponseFormatError';
        err.status = res.statusCode;
        err.headers = res.headers;
        err.data = data;
        throw err;
      }
    }
  }

  return {
    data: data,
    status: res.statusCode,
    headers: res.headers
  };
}

exports.request = request;
